source/opt/basic_block.h

```diff
@@ -106,38 +103,36 @@ class BasicBlock {
                              bool run_on_debug_line_insts = false);
 
   // Runs the given function |f| on each label id of each successor block
-  void ForEachSuccessorLabel(
-      const std::function<void(const uint32_t)>& f);
+  void ForEachSuccessorLabel(const std::function<void(const uint32_t)>& f);
 
   // Runs the given function |f| on the merge and continue label, if any
-  void ForMergeAndContinueLabel(
-      const std::function<void(const uint32_t)>& f);
+  void ForMergeAndContinueLabel(const std::function<void(const uint32_t)>& f);
 
  private:
   // The enclosing function.
   Function* function_;
   // The label starting this basic block.
   std::unique_ptr<Instruction> label_;
   // Instructions inside this basic block, but not the OpLabel.
-  std::vector<std::unique_ptr<Instruction>> insts_;
+  InstructionList insts_;
 };
 
 inline BasicBlock::BasicBlock(std::unique_ptr<Instruction> label)
     : function_(nullptr), label_(std::move(label)) {}
 
 inline void BasicBlock::AddInstruction(std::unique_ptr<Instruction> i) {
-  insts_.emplace_back(std::move(i));
+  insts_.push_back(i.release());
 }
 
 inline void BasicBlock::AddInstructions(BasicBlock* bp) {
   auto bEnd = end();
-  (void) bEnd.InsertBefore(&bp->insts_);
+  (void) bEnd.MoveBefore(&bp->insts_);
 }
 
 inline void BasicBlock::ForEachInst(const std::function<void(Instruction*)>& f,
                                     bool run_on_debug_line_insts) {
   if (label_) label_->ForEachInst(f, run_on_debug_line_insts);
-  for (auto& inst : insts_) inst->ForEachInst(f, run_on_debug_line_insts);
+  for (auto& inst : insts_) inst.ForEachInst(f, run_on_debug_line_insts);
```

Comment:
There is no "ForEachInst" implemented in the InstructionList class.  In this loop, we are calling ForEachInst on an instruction, not the instruction list.  It does this because the Instruction class contains a vector of instruction for the debug information.  We want to call it on the Instruction itself, and associated debug instructions.

Reply:
Ah, right.